packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,573 @@
+    this.persistentApis = persistentApis;
+    this.shufflingCache = shufflingCache;
+    this.getHeadState = getHeadState;
+  }
+
+  /**
+   * Reload checkpoint state keys from the last run.
+   */
+  async init(): Promise<void> {
+    const persistentKeys = await this.persistentApis.readKeys();
+    for (const persistentKey of persistentKeys) {
+      const cp = this.persistentApis.persistentKeyToCheckpoint(persistentKey);
+      this.cache.set(toCheckpointKey(cp), persistentKey);
+    }
+  }
+
+  /**
+   * Get a state from cache, it may reload from disk.
+   * This is an expensive api, should only be called in some important flows:
+   * - Validate a gossip block
+   * - Get block for processing
+   * - Regen head state
+   */
+  async getOrReload(cp: CheckpointHex): Promise<CachedBeaconStateAllForks | null> {
+    const cpKey = toCheckpointKey(cp);
+    const inMemoryState = this.get(cpKey);
+    if (inMemoryState) {
+      return inMemoryState;
+    }
+
+    const persistentKey = this.cache.get(cpKey);
+    if (persistentKey === undefined) {
+      return null;
+    }
+
+    if (!isPersistentKey(persistentKey)) {
+      // should not happen, in-memory state is handled above
+      throw new Error("Expected persistent key");
+    }
+
+    const logMeta = {persistentKey: toHexString(persistentKey)};
+
+    // reload from disk or db based on closest checkpoint
+    this.logger.verbose("Reload: read state", logMeta);
+    const newStateBytes = await this.persistentApis.read(persistentKey);
+    if (newStateBytes === null) {
+      this.logger.verbose("Reload: read state failed", logMeta);
+      return null;
+    }
+    this.logger.verbose("Reload: read state successfully", logMeta);
+    this.metrics?.stateRemoveCount.inc({reason: RemovePersistedStateReason.reload});
+    this.metrics?.stateReloadSecFromSlot.observe(this.clock?.secFromSlot(this.clock?.currentSlot ?? 0) ?? 0);
+    const closestState = this.findSeedStateToReload(cp) ?? this.getHeadState?.();
+    if (closestState == null) {
+      throw new Error("No closest state found for cp " + toCheckpointKey(cp));
+    }
+    this.metrics?.stateReloadEpochDiff.observe(Math.abs(closestState.epochCtx.epoch - cp.epoch));
+    this.logger.verbose("Reload: found closest state", {...logMeta, seedSlot: closestState.slot});
+    const timer = this.metrics?.stateReloadDuration.startTimer();
+
+    try {
+      const newCachedState = loadCachedBeaconState(closestState, newStateBytes, {
+        shufflingGetter: this.shufflingCache.getSync.bind(this.shufflingCache),
+      });
+      timer?.();
+      this.logger.verbose("Reload state successfully", {
+        ...logMeta,
+        stateSlot: newCachedState.slot,
+        seedSlot: closestState.slot,
+      });
+      // only remove persisted state once we reload successfully
+      this.persistentApis
+        .remove(persistentKey)
+        .catch((e) => this.logger.debug("Error removing persisted state", logMeta, e));
```

Comment:
Is it necessary to call remove here? Can it be scheduled to be removed in a prune step?

Reply:
yes, in that case we need to model the in-memory cache item as:
```typescript
type InMemoryCacheItem = {
  type: CacheType.inMemory;
  state: CachedBeaconStateAllForks;
  // this is set when we reload a state
  persistedKey?: PersistedKey
};
```

so that when finalized checkpoint, we prune based on the above reloaded `persistedKey`